round-based/src/simulation/sim_sync.rs

```diff
@@ -0,0 +1,222 @@
+                        }
+                        ProceedResult::NeedsOneMoreMessage => {
+                            *wants_one_more_msg = true;
+                            continue 'this_party;
+                        }
+                        ProceedResult::Output(out) => {
+                            *party_state = Party::Finished(out);
+                            parties_left -= 1;
+                            continue 'next_party;
+                        }
+                        ProceedResult::Yielded => {
+                            continue 'this_party;
+                        }
+                        ProceedResult::Error(err) => {
+                            return Err(Reason::ExecutionError(err).into());
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(self
+            .parties
+            .into_iter()
+            .map(|party| match party {
+                Party::Active { .. } => {
+                    unreachable!("there must be no active parties when `parties_left == 0`")
+                }
+                Party::Finished(out) => out,
+            })
+            .collect())
+    }
+}
+
+/// Error returned by [`SimulationSync::run`]
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct SimulationSyncError(Reason);
+
+#[derive(Debug, thiserror::Error)]
+enum Reason {
+    #[error("save incoming message")]
+    SaveIncomingMsg,
+    #[error("execution error")]
+    ExecutionError(#[source] crate::state_machine::ExecutionError),
+}
+
+impl From<Reason> for SimulationSyncError {
+    fn from(err: Reason) -> Self {
+        Self(err)
+    }
+}
+
+struct MessagesQueue<M> {
+    queue: Vec<VecDeque<Incoming<M>>>,
+    next_id: u64,
+}
+
+impl<M: Clone> MessagesQueue<M> {
+    fn new(n: usize) -> Self {
+        Self {
+            queue: vec![VecDeque::new(); n],
+            next_id: 0,
+        }
+    }
+
+    fn send_message(&mut self, sender: u16, msg: Outgoing<M>) {
+        match msg.recipient {
+            MessageDestination::AllParties => {
+                let mut msg_ids = self.next_id..;
+                for (destination, msg_id) in (0..)
+                    .zip(&mut self.queue)
+                    .filter(|(recipient_index, _)| *recipient_index != sender)
+                    .map(|(_, msg)| msg)
+                    .zip(msg_ids.by_ref())
+                {
+                    destination.push_back(Incoming {
+                        id: msg_id,
+                        sender,
+                        msg_type: MessageType::Broadcast,
+                        msg: msg.msg.clone(),
+                    })
+                }
+                self.next_id = msg_ids.next().unwrap();
+            }
+            MessageDestination::OneParty(destination) => {
+                let next_id = self.next_id;
+                self.next_id += 1;
+
+                self.queue[usize::from(destination)].push_back(Incoming {
```

Comment:
How tolerant are we to panics from index out of range? As I found yesterday, sometimes you can accidentally send a message to 0xaaaa

Reply:
Good point, we should return a proper error in such cases.

Reply:
Fixed it